### DISCUSSION TITLE
Making sense of static & dynamic rendering

### DISCUSSION THREAD

**Discussion by nareshbhatia** [vercel/next.js#54448] [App Router]
What is the definitive way of knowing if a route is statically rendered or dynamically rendered? My initial thought was that putting `console.log()` at the page level would log at build time for statically rendered routes and at runtime for dynamically rendered routes. However that doesn't appear to be the case. I am finding that dynamically rendered routes are also logging at build time. This is contrary to the [docs](https://nextjs.org/docs/app/building-your-application/caching#static-and-dynamic-rendering) which say "Static routes are cached by default, whereas dynamic routes are rendered at request time, and not cached."

### Questions
1. Why are dynamic routes being rendered at build time?
2. Is there a better way to know which routes are static vs. dynamic? Can we rely on the report printed at the end of a build (with a  `○` for static rendering and a `λ` for dynamic rendering

### Detail
Here's my [sample app](https://github.com/nareshbhatia/next13-rendering-concepts) for understanding Next.js 13 rendering concepts.

It has the following routes with expected static vs. dynamic rendering behavior:

Route | Expected Rendering Behavior
----- | -------------------------- 
/server-component-1 | Static
/server-component-2 | Static 
/server-component-3?q=top-10 | Dynamic 
/server-component-4 | Dynamic 
/server-component-5?q=top-10 | Dynamic 
/movie-magic-rsc | Dynamic
/client-component-1 | Shouldn't be static or dynamic, correct?

Here's the build report (console.logs reordered slightly for clarity):

```
> @movie-magic/movie-magic@0.0.1 build
> next build

- info Loaded env from /Users/naresh.bhatia/projects/next13-rendering-concepts/apps/movie-magic/.env
- info Creating an optimized production build
- info Compiled successfully
- info Linting and checking validity of types
- info Collecting page data
[    ] - info Generating static pages (0/12)
----> Rendering ServerComponent1
----> Rendering ServerComponent1
----> Rendering ServerComponent2
----> HTTP GET http://localhost:8080/top-10-movies
----> Rendering ServerComponent2
----> HTTP GET http://localhost:8080/top-10-movies
----> Rendering ServerComponent3
----> searchParams {}
----> Rendering ServerComponent3
----> searchParams {}
----> Rendering ServerComponent4
----> HTTP GET http://localhost:8080/top-10-movies
----> Rendering ServerComponent5
----> searchParams {}
----> Rendering ServerComponent5
----> searchParams {}
----> Rendering Movie Magic RSC
----> HTTP GET http://localhost:8080/movies
- info Generating static pages (12/12)
- info Finalizing page optimization

Route (app)                                Size     First Load JS
┌ ○ /                                      188 B          84.4 kB
├ ○ /api/hello                             0 B                0 B
├ ○ /client-component-1                    1.45 kB        85.7 kB
├ ○ /favicon.ico                           0 B                0 B
├ λ /movie-magic-rsc                       5.15 kB        89.4 kB
├ ○ /server-component-1                    188 B          84.4 kB
├ ○ /server-component-2                    188 B          84.4 kB
├ λ /server-component-3                    188 B          84.4 kB
├ λ /server-component-4                    188 B          84.4 kB
└ λ /server-component-5                    188 B          84.4 kB
+ First Load JS shared by all              78.5 kB
  ├ chunks/1dd3208c-9d03371ca40c0ae0.js    50.5 kB
  ├ chunks/575-c1e71e8e8cdf92f2.js         26.1 kB
  ├ chunks/main-app-0ad136c301e5e667.js    222 B
  └ chunks/webpack-2f1a632cffd90049.js     1.68 kB

Route (pages)                              Size     First Load JS
─ ○ /404                                   178 B          76.5 kB
+ First Load JS shared by all              76.4 kB
  ├ chunks/framework-0857c1c24dbfb1d6.js   45.1 kB
  ├ chunks/main-88481a397fb5fcf0.js        29.4 kB
  ├ chunks/pages/_app-12a7bb3ca92362ee.js  195 B
  └ chunks/webpack-2f1a632cffd90049.js     1.68 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

### Questions
1. Why are dynamic routes (`/server-component-3/4/5` & `/movie-magic-rsc`) being rendered at build time?
2. Except for `/server-component-4` & `/movie-magic-rsc`, all other routes are rendered twice, why?
3. `/server-component-4` & `/movie-magic-rsc` are dynamically rendered, so why are they fetching at build time?
4. Report correctly shows all routes as either statically or dynamically rendered, except for `/client-component-1`. This route is reported as statically rendered, which is confusing because this page is a client component and my understanding is that  static and dynamic rendering apply only to server components, correct?

**Comment by devjiwonchoi** [comment]
Note that this is a partial answer, will investigate and update it afterwards.

Based on my knowledge, all "dynamic" pages are built on build time limited to the scripts for the actual rendering on request.

The difference is that they do not build static files such as `html`, and I believe if you open up the `.next` outDir folder, there won't be any HTML files for server-component3~5 files.

### TL;DR

The static files such as `html`, and `meta` won't be generated at build time, but `console.log` will be logged since Next.js needs to build script files for dynamic rendering on production.

P.S. Thank you for sharing your great observation, it is definitely inspiring.

**Reply by mikhkosvn** [reply]
still seeking for an answer?
